### PR TITLE
automated test to validate auto image pull

### PR DESCRIPTION
**Pull request by runewake2**
This adds an automated test that verifies images are automatically pulled when kpt functions are executed.

### CODE REVIEW

**Comment by natasha41575** [comment]
This LGTM but should wait for lead approval

@droot @frankfarzan This is for line 11 of fn-render in go/kpt-test-cases

**Review by Shell32-Natsu — COMMENTED** [review]
Thanks for this test. Maybe you can update the implementation here
https://github.com/GoogleContainerTools/kpt/blob/next/internal/fnruntime/container.go#L146-L156 
to use `docker image inspect` instead of `docker image ls`.

**Comment by runewake2** [comment]
Is that the behavior we want @Shell32-Natsu? The commands behave slightly different. `docker image ls` will succeed if you do not provide a tag for example but inspect may fail.

```
$ docker image ls gcr.io/kpt-fn/search-replace
REPOSITORY                     TAG                 IMAGE ID            CREATED             SIZE
gcr.io/kpt-fn/search-replace   v0.1                187c912bbdb5        2 weeks ago         19MB

$ docker image inspect gcr.io/kpt-fn/search-replace                                                                                      
Error: No such image: gcr.io/kpt-fn/search-replace
```

**Comment by Shell32-Natsu** [comment]
> Is that the behavior we want @Shell32-Natsu? The commands behave slightly different. `docker image ls` will succeed if you do not provide a tag for example but inspect may fail.
> 
> ```
> $ docker image ls gcr.io/kpt-fn/search-replace
> REPOSITORY                     TAG                 IMAGE ID            CREATED             SIZE
> gcr.io/kpt-fn/search-replace   v0.1                187c912bbdb5        2 weeks ago         19MB
> 
> $ docker image inspect gcr.io/kpt-fn/search-replace                                                                                      
> Error: No such image: gcr.io/kpt-fn/search-replace
> ```

I think this is the expected. For our functions in the catalog, they don't have `latest` tag so pulling it without tag is not possible. In most cases user needs to provide the image name as well as tag.

**Comment by runewake2** [comment]
I'll take a look, but I don't think that change is straightforward. `docker image ls random-image` returns successfully but `inspect` would fail. We would need to update our logic to make this behave how we expect.

File: internal/fnruntime/container_test.go
Comment (droot) [inline internal/fnruntime/container_test.go:108]: It's not clear to me what exactly are we testing here ?

Comment (droot) [reply internal/fnruntime/container_test.go:108]: When you invoke `instance.Run()`:
1. Image could be pulled using the `prepareImage()`
2. Or it could be pulled using the `docker run`

so either case, `docker image inspect` will succeed.. right ? or may be I am missing something.

Comment (runewake2) [reply internal/fnruntime/container_test.go:108]: This is specifically targeting the container functions and that executing the container Command will pull the image. It is not testing **how** that image is pulled (yet). This is meant as a proxy for manually testing this functionality.

`docker image inspect` exits with non-zero exit code if the image does not exist. A quick example:

```
$ docker pull postgres:latest
latest: Pulling from library/postgres
Status: Downloaded newer image for postgres:latest
docker.io/library/postgres:latest

$ docker image rm postgres:latest
Untagged: postgres:latest
...

$ docker image inspect postgres:latest
[]
Error: No such image: postgres:latest
```

This test is meant to:

1. Remove any existing local image reference (if one exists)
2. Execute the function as a no-op
3. Validate that the function image is available locally

The test case I had in mind was:

> Given an environment that does not have a function image locally, when kpt executes that function, then the function image is checked out locally.

This could be refactored to target specifically covering `prepareImage` if we'd prefer but is currently broader than that. If you'd like I could append a test to this PR that introduces that coverage.

Comment (droot) [reply internal/fnruntime/container_test.go:108]: > This could be refactored to target specifically covering prepareImage if we'd prefer but is currently broader than that. If you'd like I could append a test to this PR that introduces that coverage.

No, not in favor of testing internal functions. Infact, I would like the test case to be:

> Given an environment that does not have a function image locally, when kpt _should be able to execute_ that function successfully.

Ideally having it in e2e test case would be better but we don't have a way to setup the test-case today where we ensure the function image isn't present before running the test. 

But this is fine for now.

**Review by droot — APPROVED** [review]

File: internal/fnruntime/container_test.go
Comment (droot) [reply internal/fnruntime/container_test.go:108]: Today @Shell32-Natsu added support for running `setup.sh` script in e2e test framework https://github.com/GoogleContainerTools/kpt/pull/2032

So, we can implement this test as e2e test. That will be better. So copy one of the simple tests and add `setup.sh` that deletes the docker images being used in that test and successful run of the test will achieve the same outcome as here.

**Review by droot — CHANGES_REQUESTED** [review]
Suggest to implement this as e2e test since support for `setup.sh` and `teardown.sh` has been added.

File: internal/fnruntime/container_test.go
Comment (frankfarzan) [inline internal/fnruntime/container_test.go:19]: @droot are these supposed to be unit tests or e2e tests?

Comment (droot) [reply internal/fnruntime/container_test.go:19]: Actually these were added before we had `e2e` tests. I have suggested @runewake2 to implement this test in `e2e` now because with `setup.sh` support in e2e harness. We can safely remove these tests after that is done. @Shell32-Natsu Can you pl. confirm.

**Review by droot — APPROVED** [review]
Pl. update the row corresponding to this test green in the test sheet and also, mark it green.

Thank you!

**Pull request merged**
